File: api-doclet/src/main/java/org/conscrypt/doclet/FilterDoclet.java
/*
 * Copyright (C) 2010 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/*
 * Originally from Doclava project at
 * https://android.googlesource.com/platform/external/doclava/+/master/src/com/google/doclava/Doclava.java
 */

package org.conscrypt.doclet;

import com.sun.javadoc.*;
import com.sun.tools.doclets.standard.Standard;
import com.sun.tools.javadoc.Main;
import java.io.FileNotFoundException;
import java.lang.reflect.Array;
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;
import java.util.ArrayList;
import java.util.List;

/**
 * This Doclet filters out all classes, methods, fields, etc. that have the {@code @Internal}
 * annotation on them.
 */
public class FilterDoclet extends com.sun.tools.doclets.standard.Standard {
    public static void main(String[] args) throws FileNotFoundException {
        String name = FilterDoclet.class.getName();
        Main.execute(name, args);
    }

    public static boolean start(RootDoc rootDoc) {
        return Standard.start((RootDoc) filterHidden(rootDoc, RootDoc.class));
    }

    /**
     * Returns true if the given element has an @Internal annotation.
     */
    private static boolean hasHideAnnotation(ProgramElementDoc doc) {
        for (AnnotationDesc ann : doc.annotations()) {
            if (ann.annotationType().qualifiedTypeName().equals("org.conscrypt.Internal")) {
                return true;
            }
        }
        return false;
    }

    /**
     * Returns true if the given element is hidden.
     */
    private static boolean isHidden(Doc doc) {
        // Methods, fields, constructors.
        if (doc instanceof MemberDoc) {
            return hasHideAnnotation((MemberDoc) doc);
        }
        // Classes, interfaces, enums, annotation types.
        if (doc instanceof ClassDoc) {
            // Check the class doc and containing class docs if this is a
            // nested class.
            ClassDoc current = (ClassDoc) doc;
            do {
                if (hasHideAnnotation(current)) {
                    return true;
                }
                current = current.containingClass();
            } while (current != null);
        }
        return false;
    }

    /**
     * Filters out hidden elements.
     */
    private static Object filterHidden(Object o, Class<?> expected) {
        if (o == null) {
            return null;
        }

        Class<?> type = o.getClass();
        if (type.getName().startsWith("com.sun.")) {
            // TODO: Implement interfaces from superclasses, too.
            return Proxy.newProxyInstance(
                    type.getClassLoader(), type.getInterfaces(), new HideHandler(o));
        } else if (o instanceof Object[]) {
            Class<?> componentType = expected.getComponentType();
            if (componentType == null) {
                return o;
            }

            Object[] array = (Object[]) o;
            List<Object> list = new ArrayList<Object>(array.length);
            for (Object entry : array) {
                if ((entry instanceof Doc) && isHidden((Doc) entry)) {
                    continue;
                }
                list.add(filterHidden(entry, componentType));
            }
            return list.toArray((Object[]) Array.newInstance(componentType, list.size()));
        } else {
            return o;
        }
    }

    /**
     * Filters hidden elements.
     */
    private static class HideHandler implements InvocationHandler {
        private final Object target;

        public HideHandler(Object target) {
            this.target = target;
        }

        @Override
        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
            String methodName = method.getName();
            if (args != null) {
                if (methodName.equals("compareTo") || methodName.equals("equals")
                        || methodName.equals("overrides") || methodName.equals("subclassOf")) {
                    args[0] = unwrap(args[0]);
                }
            }

            try {
                return filterHidden(method.invoke(target, args), method.getReturnType());
            } catch (InvocationTargetException e) {
                e.printStackTrace();
                throw e.getTargetException();
            }
        }

        private static Object unwrap(Object proxy) {
            if (proxy instanceof Proxy)
                return ((HideHandler) Proxy.getInvocationHandler(proxy)).target;
            return proxy;
        }
    }
}

File: api-doclet/src/main/kotlin/org/conscrypt/doclet/FilterDoclet.kt
/*
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.conscrypt.doclet

import com.sun.source.util.DocTrees
import jdk.javadoc.doclet.Doclet
import jdk.javadoc.doclet.DocletEnvironment
import jdk.javadoc.doclet.Reporter
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.Locale
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.util.Elements
import javax.lang.model.util.Types

/**
 * A Doclet which can filter out internal APIs in various ways and then render the results
 * as HTML.
 *
 * See also: The Element.isFiltered extension function below to see what is filtered.
 */
class FilterDoclet : Doclet {
    companion object {
        lateinit var docTrees: DocTrees
        lateinit var elementUtils: Elements
        lateinit var typeUtils: Types
        lateinit var outputPath: Path
        lateinit var cssPath: Path
        var baseUrl: String = "https://docs.oracle.com/en/java/javase/21/docs/api/java.base/"
        const val CSS_FILENAME = "styles.css"
        var outputDir = "."
        var docTitle = "DOC TITLE"
        var windowTitle = "WINDOW TITLE"
        var noTimestamp: Boolean = false
        val classIndex = ClassIndex()
    }

    override fun init(locale: Locale?, reporter: Reporter?) = Unit // TODO
    override fun getName() = "FilterDoclet"
    override fun getSupportedSourceVersion(): SourceVersion = SourceVersion.latest()

    override fun run(environment: DocletEnvironment): Boolean {
        docTrees = environment.docTrees
        elementUtils = environment.elementUtils
        typeUtils = environment.typeUtils
        outputPath = Paths.get(outputDir)
        cssPath = outputPath.resolve(CSS_FILENAME)
        Files.createDirectories(outputPath)

        classIndex.addVisible(environment.includedElements)

        try {
            generateClassFiles()
            generateIndex()
            return true
        } catch (e: Exception) {
            System.err.println("Error generating documentation: " + e.message)
            e.printStackTrace()
            return false
        }
    }

    private fun generateClassFiles() = classIndex.classes().forEach(::generateClassFile)

    private fun generateIndex() {
        val indexPath = outputPath.resolve("index.html")

        html {
            body(
                title = docTitle,
                stylesheet = relativePath(indexPath, cssPath),
            ) {
                div("index-container") {
                    h1(docTitle, "index-title")
                    compose {
                        classIndex.generateHtml()
                    }
                }
            }
        }.let {
            Files.newBufferedWriter(indexPath).use { writer ->
                writer.write(it)
            }
        }
    }

    private fun generateClassFile(classInfo: ClassInfo) {
        val classFilePath = outputPath.resolve(classInfo.fileName)
        Files.createDirectories(classFilePath.parent)
        val name = classInfo.innerName()

        html {
            body(
                title = "$name - Conscrypt API",
                stylesheet = relativePath(classFilePath, cssPath),
            ) {
                compose {
                    classInfo.generateHtml()
                }
            }
        }.let {
            Files.newBufferedWriter(classFilePath).use { writer ->
                writer.write(it)
            }
        }
    }

    private fun relativePath(from: Path, to: Path) = from.parent.relativize(to).toString()

    override fun getSupportedOptions(): Set<Doclet.Option> {
        return setOf<Doclet.Option>(
            StringOption(
                "-d",
                "<directory>",
                "Destination directory for output files"
            ) { d: String -> outputDir = d },
            StringOption(
                "-doctitle",
                "<title>",
                "Document title"
            ) { t: String -> docTitle = t },
            StringOption(
                "-windowtitle",
                "<title>",
                "Window title"
            ) { w: String -> windowTitle = w },
            StringOption(
                "-link",
                "<link>",
                "Link"
            ) { l: String -> baseUrl = l },
            BooleanOption(
                "-notimestamp",
                "Something"
            ) { noTimestamp = true })
    }
}

// Called to determine whether to filter each public API element.
fun Element.isFiltered() =
        hasJavadocTag("hide") || hasAnnotation("org.conscrypt.Internal")
